tests/transport/spdp/spdp_transport.cpp

```diff
@@ -0,0 +1,485 @@
+      char subm = recv_mb_.rd_ptr()[0], flags = recv_mb_.rd_ptr()[1];
+      ser.swap_bytes((flags & FLAG_E) != ACE_CDR_BYTE_ORDER);
+      switch (subm) {
+      case DATA:
+        if (!recv_data(ser, peer)) return false;
+        break;
+      default:
+#ifdef OPENDDS_NO_CONTENT_SUBSCRIPTION_PROFILE
+        ACE_DEBUG((LM_INFO, "Received submessage type: %u\n", unsigned(subm)));
+#else
+        if (static_cast<size_t>(subm) < gen_OpenDDS_RTPS_SubmessageKind_names_size) {
+          ACE_DEBUG((LM_INFO, "Received submessage type: %C\n",
+                     gen_OpenDDS_RTPS_SubmessageKind_names[static_cast<size_t>(subm)]));
+        } else {
+          ACE_ERROR((LM_ERROR, "ERROR: Received unknown submessage type: %u\n",
+                     unsigned(subm)));
+        }
+#endif
+        SubmessageHeader smh;
+        if (!(ser >> smh)) {
+          ACE_ERROR((LM_ERROR, "ERROR: in handle_input() failed to deserialize "
+                               "SubmessageHeader\n"));
+          return -1;
+        }
+        if (smh.submessageLength == 0) {
+          return 0;
+        }
+        recv_mb_.rd_ptr(smh.submessageLength);
+      }
+    }
+    return 0;
+  }
+
+  bool recv_data(Serializer& ser, const ACE_INET_Addr&)
+  {
+    DataSubmessage data;
+    if (!(ser >> data)) {
+      ACE_ERROR((LM_ERROR,
+        "ERROR: recv_data() failed to deserialize DataSubmessage\n"));
+      return false;
+    }
+    ACE_DEBUG((LM_INFO, "recv_data() seq = %d\n", data.writerSN.low));
+    if (data.smHeader.submessageLength) {
+      ser.skip(data.smHeader.submessageLength - 20);
+      // 20 == size of Data headers after smHeader (assuming no Inline QoS)
+    } else {
+      ser.skip(8);  // our data payloads are 8 bytes
+    }
+    return true;
+  }
+
+  ACE_SOCK_Dgram& sock_;
+  Header hdr_, recv_hdr_;
+  ACE_Message_Block recv_mb_;
+};
+
+
+void reactor_wait()
+{
+  ACE_Time_Value half_second(0, ACE_ONE_SECOND_IN_USECS / 2);
+  ACE_Reactor::instance()->run_reactor_event_loop(half_second);
+}
+
+struct ReactorTask : ACE_Task_Base {
+
+  ReactorTask()
+  {
+    activate();
+  }
+
+  int svc()
+  {
+    ACE_Reactor* reactor = ACE_Reactor::instance();
+    ACE_thread_t old_owner;
+    reactor->owner(ACE_Thread_Manager::instance()->thr_self(), &old_owner);
+    reactor_wait();
+    reactor->owner(old_owner);
+    return 0;
+  }
+};
+
+bool run_test()
+{
+  // Create a "test participant" which will use sockets directly
+  // This will act like a remote participant.
+  ACE_SOCK_Dgram test_part_sock;
+  ACE_INET_Addr test_part_addr;
+  if (!open_appropriate_socket_type(test_part_sock, test_part_addr)) {
+    ACE_ERROR((LM_ERROR, ACE_TEXT("ERROR: run_test() unable to open test_part_sock\n")));
+    return false;
+  }
+  test_part_sock.get_local_addr(test_part_addr);
+  test_part_addr.set(test_part_addr.get_port_number(),
+#ifdef OPENDDS_SAFETY_PROFILE
+    "127.0.0.1"
+#else
+    "localhost"
+#endif
+    );
+
+  ACE_INET_Addr send_addr("239.255.0.1:7400");
```

Comment:
Need to determine if we should be changing port numbers (see OPENDDS_RTPS_DEFAULT_D0) or binding to loopback only, etc., to make sure two concurrent tests don't conflict.